### ISSUE TITLE
Digital Ocean "Waiting for SSH..." seems to hang a lot

### ISSUE THREAD

**Issue by cjellick** [docker/machine#951]
We see a lot of problems during the Waiting for SSH phase of provisioning a Digital Ocean machine.

Here's one instance of the command and resulting error. Forgive/ignore the weird formatting. We're calling docker-machine from a golang program via `exec.Cmd`

```
docker-machine create -d digitalocean --digitalocean-image ubuntu-14-04-x64 --digitalocean-size 1gb --digitalocean-region nyc3 --digitalocean-access-token <redacted> foo2
```
```
stdout: time=\"2015-04-02T20:59:42Z\" level=\"info\" msg=\"Creating CA: /var/lib/cattle/machine/fa34d3ad-de89-4f86-9fe0-ed12ed8ca77f/certs/ca.pem\" " 
stdout: time=\"2015-04-02T20:59:43Z\" level=\"info\" msg=\"Creating client certificate: /var/lib/cattle/machine/fa34d3ad-de89-4f86-9fe0-ed12ed8ca77f/certs/cert.pem\" "
stdout: time=\"2015-04-02T20:59:44Z\" level=\"info\" msg=\"Creating SSH key...\" "
stdout: time=\"2015-04-02T20:59:45Z\" level=\"info\" msg=\"Creating Digital Ocean droplet...\" "
stdout: time=\"2015-04-02T20:59:46Z\" level=\"info\" msg=\"Waiting for SSH...\" "
stdout: time=\"2015-04-02T21:04:48Z\" level=\"error\" msg=\"Error creating machine: exit status 255\" "
stdout: time=\"2015-04-02T21:04:48Z\" level=\"warning\" msg=\"You will want to check the provider to make sure the machine and associated resources were properly removed.\" "
stdout: time=\"2015-04-02T21:04:48Z\" level=\"fatal\" msg=\"Error creating machine\" "
```

**Comment by ehazlett** [comment]
Is this on the latest rc3?

**Comment by cjellick** [comment]
yup.
```
md5sum /usr/bin/docker-machine
04b7361d1dd1afdfd8e282da77fa2cf4  /usr/bin/docker-machine
```

**Comment by ehazlett** [comment]
thanks! i'm noticing the same issues now -- digging

**Comment by UserTaken** [comment]
I have also experienced hanging in all releases when creating multiple droplets in rapid succession. A somewhat related problem I've been having with Digital Ocean is the lack of swap as described in [#829](https://github.com/docker/machine/issues/829) so I'm wondering if something can be done about that?

**Comment by ehazlett** [comment]
@cjellick could you try fix in #952 to see if it works any better for you?  I haven't had any issues after that but it may just be coincidence :)

**Comment by cjellick** [comment]
Will do. Thanks!

**Comment by themattrix** [comment]
I've encountered this too.

Additionally, it often gets stuck on the next step, "Configuring Machine...". It exits with an error after a few minutes.

```
INFO[0000] Creating SSH key...
INFO[0000] Creating Digital Ocean droplet...
INFO[0003] Waiting for SSH...
INFO[0130] Configuring Machine...
ERRO[0257] Error creating machine: exit status 255
WARN[0257] You will want to check the provider to make sure the machine and associated resources were properly removed.
FATA[0257] Error creating machine
```

The actual VM seems to have been created normally and is accessible over SSH.

**Comment by ehazlett** [comment]
@themattrix what version of machine are you using?  can you run with `-D` for debug?

**Comment by clnperez** [comment]
I'm seeing this as well, having built from upstream: 
$ machine -v
$ machine version 0.2.0 (HEAD)

It was a pretty long (maybe an hour? not sure) lag yesterday, but when I checked this morning it had finished. I just ran again with -D and it finished in two minutes. I ran again without -D and it hung. Odd, so I ran again with -D, and it hung:

time="2015-04-13T10:44:50-05:00" level="info" msg="Creating SSH key..." 
time="2015-04-13T10:44:50-05:00" level="debug" msg="executing: /usr/bin/ssh-keygen ssh-keygen -t rsa -N  -f /home/user/.docker/machine/machines/dovm1/id
_rsa\n" 
Generating public/private rsa key pair.
Your identification has been saved in /home/user/.docker/machine/machines/dovm1/id_rsa.
Your public key has been saved in /home/user/.docker/machine/machines/dovm1/id_rsa.pub.
The key fingerprint is:
[deleted] user@system
The key's randomart image is:
+---[RSA 2048]----+
[deleted]

time="2015-04-13T10:44:51-05:00" level="info" msg="Creating Digital Ocean droplet..." 
time="2015-04-13T10:44:53-05:00" level="debug" msg="Created droplet ID 4847925, IP address 104.131.160.180" 
time="2015-04-13T10:44:53-05:00" level="debug" msg="Getting to WaitForSSH function..."

**Comment by ehazlett** [comment]
Hmm that looked like it finished ok.  I have noticed some issues occasionally with DO but they usually timeout well before an hour.

**Comment by themattrix** [comment]
@ehazlett I've been using docker-machine `0.1.0`. This issue only happens occasionally, but when it does it usually occurs several times in a row. I've enabled debug mode and will report back when I see it again.

**Comment by ehazlett** [comment]
@themattrix thanks.  could you also try it with the latest RC?  several fixes have been made.  thanks!

**Comment by themattrix** [comment]
@ehazlett sure thing!

**Comment by stefan-jones** [comment]
I've seen this with Docker 1.6.2, docker-machine 0.30, docker-compose 1.2

Four VMs with Centos Linux  2.6.32-504.16.2.el6.x86_64.

Created a swarm on node 1 of a four node cluster. Each attempt to create a machine hung at "Importing SSH key...."

Command line:

[root@XXXX-vm-1 ~]# docker-machine create -d generic --generic-ip-address FOO --generic-ssh-key ~/.ssh/id_rsa --swarm --swarm-master --swarm-discovery token://$TOKEN --engine-label varnish=true XXXX-vm-1 
Creating CA: /root/.docker/machine/certs/ca.pem
Creating client certificate: /root/.docker/machine/certs/cert.pem
Importing SSH key...

I ^C'd out of the hung session after a half hour: Inspection of the directories under ~/.docker showed that the key had been imported, and a machine had been created albeit stopped.

[root@XXXX-vm-1 ~]#  ls -l ~/.docker/machine/machines
total 4
drwx------ 2 root root 4096 May 29 09:08 XXXX-vm-1
[root@XXXX-vm-1 ~]#  ls -l /root/.docker/machine/machines/XXXX-vm-1/
total 8
-rw------- 1 root root 1609 May 29 09:08 config.json
-rw------- 1 root root 1743 May 29 09:08 id_rsa

[root@XXXX-vm-1 ~]# docker-machine ls
NAME            ACTIVE   DRIVER    STATE     URL                      SWARM
XXXX-vm-1            generic   Stopped   tcp://FOO:2376   XXXX-vm-1 (master)

I see folks have posted log entries up above. Which log are these from? I don't see anything in /var/log/docker as verbose as that posted up above.

**Comment by ehazlett** [comment]
@stefan-jones could you try running with `docker-machine -D create ...` for some debug?  Thanks!

**Comment by stefan-jones** [comment]
I removed the existing machines, and entered the same command line, this time with the –D debug flag.

An endless stream of “Testing TCP connection to: [ip address of the host system]:22”

That is , I’m running the command on system with IP address x.x.x.x, and it is testing a TCP connection to itself, port 22.

I’ve heard of Docker occasionally mucking up iptables.

Stefan Jones
ARRIS

o: +1 (503) 495-9458
c: -

**Comment by stefan-jones** [comment]
OK, I'm going to pound my head against a desk for a bit. I recently started using a new set of VMs and got an IP address wrong when composing the command line. Using the correct IP address got me to teh point where I'm being asked for the key passphrase.

I do wonder if the the connection test should have timed out after a while.

**Comment by ehazlett** [comment]
@stefan-jones does your key have a passphrase?   the ssh user also needs passwordless sudo.

**Comment by stefan-jones** [comment]
While there are still some issues with the TCP connection that we are working through, the long hang was due to my error. I bungled a search and replace to update IP addresses in the file of command lines I was cutting and pasting. The IP address it was looking for did not exist in our network.

Once I corrected the IP address the procedure got to the point where it asked about the password. There were issues after that that our developers are looking into (the command as entered worked on a Ubuntu box but not my Centos based cluster) but that is not relevant to this bug.

The long hang might be a justification for implementing a timeout; the machine create routine seemed happy to try to test the connection to the nonexistent address forever. However, that is an enhancement and not relevant to this issue.

Stefan Jones
ARRIS

o: +1 (503) 495-9458
c: -
